From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Gwan-gyeong Mun <gwan-gyeong.mun@intel.com>
Date: Mon, 11 Nov 2024 14:07:18 +0200
Subject: [PATCH] drm/xe/eudebug: Enable EU pagefault handling

The XE2 (and PVC) HW has a limitation that the pagefault due to invalid
access will halt the corresponding EUs. To solve this problem, enable
EU pagefault handling functionality, which allows to unhalt pagefaulted
eu threads and to EU debugger to get inform about the eu attentions state
of EU threads during execution.

If a pagefault occurs, send the DRM_XE_EUDEBUG_EVENT_PAGEFAULT event
after handling the pagefault.

The pagefault handling is a mechanism that allows a stalled EU thread to
enter SIP mode by installing a temporal null page to the page table entry
where the pagefault happened.

A brief description of the page fault handling mechanism flow between KMD
and the eu thread is as follows

(1) eu thread accesses unallocated address
(2) pagefault happens and eu thread stalls
(3) XE kmd set an force eu thread exception to allow the running eu thread
    to enter SIP mode (kmd set ForceException bit of TD_CTL register)
    Not stalled (none-pagefaulted) eu threads enter SIP mode
(4) XE kmd installs temporal null page to the pagetable entry of the
    address where pagefault happened.
(5) XE kmd replies pagefault successful message to GUC
(6) When a given pagefault is fully handled temporary NULL VMA is
    removed.
(7) If multiple threads caused a pagefault and this is reflected in
    pagefault queue or worker's cache then the first pagefault
    finalization is postponed but the worker is released
(8) All subsequent pagefaults for a given VM are handled by eudebug by
    inserting NULL VMA, ACK and then remove the temporary VMA
(9) stalled eu thread resumes as per pagefault condition has resolved
(10) resumed eu thread enters SIP mode due to force exception set
     by (3)
(9) When there are no more pagefault to process the first pagefault
    finalization is performed and event is send.

As designed this feature to only work when eudbug is enabled, it should
have no impact to regular recoverable pagefault code path.

v2: - pf->q holds the vm ref so drop it (Mika)
    - streamline uapi (Mika)
    - cleanup the pagefault through producer if (Mika)
v3: - pagefault rework (Maciej)

Assisted-by: GitHub-Copilot:claude-opus-5
Cc: Matthew Brost <matthew.brost@intel.com>
Cc: Gustavo Sousa <gustavo.sousa@intel.com>
Signed-off-by: Gwan-gyeong Mun <gwan-gyeong.mun@intel.com>
Signed-off-by: Maciej Patelczyk <maciej.patelczyk@intel.com>
Signed-off-by: Mika Kuoppala <mika.kuoppala@linux.intel.com>
(cherry-picked from commit 43470d196b5935ce60783c5b688d012da5fb583e eudebug-v3-xe3p-prelim)
Signed-off-by: Bommu Krishnaiah <krishnaiah.bommu@intel.com>
---
 drivers/gpu/drm/xe/xe_device_types.h      |   3 +
 drivers/gpu/drm/xe/xe_eudebug.c           |   8 +
 drivers/gpu/drm/xe/xe_eudebug_pagefault.c | 170 +++++++++++++++++++++-
 drivers/gpu/drm/xe/xe_eudebug_pagefault.h |  21 ++-
 drivers/gpu/drm/xe/xe_eudebug_types.h     |   5 +
 drivers/gpu/drm/xe/xe_gt.c                |   2 +
 drivers/gpu/drm/xe/xe_gt_types.h          |  12 ++
 drivers/gpu/drm/xe/xe_guc_pagefault.c     |  10 +-
 drivers/gpu/drm/xe/xe_pagefault.c         | 104 ++++++++++++-
 drivers/gpu/drm/xe/xe_pagefault_types.h   |   4 +-
 10 files changed, 319 insertions(+), 20 deletions(-)

diff --git a/drivers/gpu/drm/xe/xe_device_types.h b/drivers/gpu/drm/xe/xe_device_types.h
index e261fd915..42bde484f 100644
--- a/drivers/gpu/drm/xe/xe_device_types.h
+++ b/drivers/gpu/drm/xe/xe_device_types.h
@@ -660,6 +660,9 @@ struct xe_device {
 
 		/** @eudebug.attention_scan: attention scan worker */
 		struct delayed_work attention_scan;
+
+		/** @eudebug.pagefault_reset_wq: used when there is a GT reset during pagefault */
+		struct workqueue_struct *pagefault_reset_wq;
 	} eudebug;
 #endif
 
diff --git a/drivers/gpu/drm/xe/xe_eudebug.c b/drivers/gpu/drm/xe/xe_eudebug.c
index d9cf00d97..66ac30fef 100644
--- a/drivers/gpu/drm/xe/xe_eudebug.c
+++ b/drivers/gpu/drm/xe/xe_eudebug.c
@@ -2427,6 +2427,14 @@ void xe_eudebug_init(struct xe_device *xe)
 		return;
 	}
 
+	xe->eudebug.pagefault_reset_wq = drmm_alloc_ordered_workqueue(&xe->drm,
+								      "xe-eudebug-pf-gt-reset",
+								      0);
+	if (IS_ERR(xe->eudebug.pagefault_reset_wq)) {
+		destroy_workqueue(xe->eudebug.ordered_wq);
+		return;
+	}
+
 	ret = sysfs_create_file(&xe->drm.dev->kobj, &dev_attr_enable_eudebug.attr);
 	if (ret)
 		drm_warn(&xe->drm, "eudebug sysfs init failed: %d, debugger unavailable\n", ret);
diff --git a/drivers/gpu/drm/xe/xe_eudebug_pagefault.c b/drivers/gpu/drm/xe/xe_eudebug_pagefault.c
index 9db368541..4a2b68fc6 100644
--- a/drivers/gpu/drm/xe/xe_eudebug_pagefault.c
+++ b/drivers/gpu/drm/xe/xe_eudebug_pagefault.c
@@ -17,6 +17,7 @@
 #include "xe_gt_mcr.h"
 #include "xe_gt_printk.h"
 #include "regs/xe_gt_regs.h"
+#include "xe_pm.h"
 #include "xe_vm.h"
 #include "xe_guc.h"
 #include "xe_guc_submit.h"
@@ -121,18 +122,36 @@ void xe_eudebug_pagefault_set_private(struct xe_pagefault *pf,
 				      struct xe_vm *vm)
 {
 	struct xe_eudebug_pagefault *epf;
-	struct xe_eudebug *d;
+	struct xe_eudebug *d = NULL;
+	struct xe_vm *lvm = NULL;
 
 	if (!xe_eudebug_is_enabled(gt_to_xe(pf->gt)))
 		return;
 
+	/*
+	 * This services the case when function is called on
+	 * cache hit. The xe_pagefault_queue_work() jumps to
+	 * ack_fault bypassing xe_pagefault_service().
+	 */
+	if (!vm) {
+		lvm = xe_device_asid_to_vm(gt_to_xe(pf->gt), PF_ASID(pf));
+		if (IS_ERR(lvm))
+			return;
+
+		if (!xe_vm_in_fault_mode(lvm)) {
+			xe_vm_put(lvm);
+			return;
+		}
+		vm = lvm;
+	}
+
 	/* Could be a retry from -EAGAIN handler error */
 	if ((u64)pf->producer.private & XE_EUDEBUG_PAGEFAULT_PRIVATE_EUDEBUG)
-		return;
+		goto out;
 
 	d = xe_eudebug_get_nolock_with_discovery(vm->xef);
 	if (!d)
-		return;
+		goto out;
 
 	mutex_lock(&d->eu_lock);
 	epf = d->pf.first;
@@ -146,7 +165,12 @@ void xe_eudebug_pagefault_set_private(struct xe_pagefault *pf,
 	if (epf)
 		pagefault_set_private(pf, epf);
 
-	xe_eudebug_put(d);
+out:
+	if (d)
+		xe_eudebug_put(d);
+
+	if (lvm)
+		xe_vm_put(lvm);
 }
 
 void *xe_eudebug_pagefault_get_private(void *private)
@@ -225,9 +249,25 @@ static int __clear_td_ctl_fee(struct xe_gt *gt, struct xe_eudebug *d)
 	return 0;
 }
 
+static void
+eudebug_try_svm_remap(struct xe_vm *vm, struct xe_vma *vma)
+{
+	if (xe_vm_find_overlapping_vma(vm,
+				       xe_vma_start(vma),
+				       xe_vma_size(vma))) {
+		xe_vm_destroy_vma(vma);
+		return;
+	}
+
+	if (xe_vm_insert_vma(vm, vma)) {
+		xe_vm_destroy_vma(vma);
+		xe_vm_kill(vm, true);
+	}
+}
 
 int
-xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_pagefault *pf)
+xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_vma *vma,
+			   struct xe_pagefault *pf)
 {
 	struct pagefault_fence *pf_fence;
 	struct xe_eudebug_pagefault *epf;
@@ -258,6 +298,9 @@ xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_pagefault *pf)
 	if (!xe_exec_queue_is_debuggable(q))
 		goto err_put_exec_queue;
 
+	if (vma && !xe_vma_is_cpu_addr_mirror(vma))
+		goto err_put_exec_queue;
+
 	/**
 	 * Check if there is an active pagefault.
 	 * If so, attach original epf to current pagefault and leave.
@@ -277,7 +320,8 @@ xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_pagefault *pf)
 			mutex_unlock(&d->eu_lock);
 			goto err_put_exec_queue;
 		}
-		d->pf.active++;
+		if (!pf_tagged(pf))
+			d->pf.active++;
 		INIT_LIST_HEAD(&epf->link);
 		goto out_set_epf;
 	}
@@ -342,7 +386,18 @@ out_set_epf:
 	epf->fault.type_level = pf->consumer.fault_type_level;
 	epf->fault.access_type = pf->consumer.access_type;
 
+	/*
+	 * VMA is only present on SVM pf and that pf
+	 * was early tagged (first pass).
+	 * Clear early tag and restore original private
+	 * field so it can be safely tagged again.
+	 */
+	if (vma)
+		pf->producer.private =
+			xe_eudebug_pagefault_get_private(pf->producer.private);
+
 	pagefault_set_private(pf, epf);
+	epf->svm_vma = vma;
 
 	return 0;
 
@@ -430,8 +485,13 @@ static void epf_release_vma(struct xe_eudebug_pagefault *epf)
 	struct xe_vm *vm = epf_to_vm(epf);
 	struct xe_vma *vma;
 
-	if (xe_vm_is_closed(vm))
+	if (xe_vm_is_closed(vm)) {
+		if (epf->svm_vma) {
+			xe_vm_destroy_vma(epf->svm_vma);
+			epf->svm_vma = NULL;
+		}
 		return;
+	}
 
 	if (epf->null_vma_addr) {
 		u32 psize = vm->flags & XE_VM_FLAG_64K ? SZ_64K : SZ_4K;
@@ -450,6 +510,12 @@ static void epf_release_vma(struct xe_eudebug_pagefault *epf)
 			eudebug_destroy_vma(vm, vma);
 		epf->null_vma_addr = 0;
 	}
+
+	if (epf->svm_vma) {
+		lockdep_assert_held_write(&vm->lock);
+		eudebug_try_svm_remap(epf->q->vm, epf->svm_vma);
+		epf->svm_vma = NULL;
+	}
 }
 
 static void destroy_pagefault(struct xe_eudebug_pagefault *epf)
@@ -870,6 +936,96 @@ void xe_eudebug_pagefault_signal(struct xe_eudebug *d)
 	}
 }
 
+static void eudebug_pagefault_gt_reset_work(struct work_struct *work)
+{
+	struct xe_gt *gt = container_of(work, typeof(*gt),
+					eudebug.gt_reset_work);
+	struct xe_eudebug_pagefault *epf = NULL;
+	struct xe_device *xe = gt_to_xe(gt);
+	struct xe_eudebug *d = NULL;
+	struct xe_vm *vm;
+	unsigned long i;
+
+	/*
+	 * Search for deferred abandoned eudebug pagefault.
+	 * xe_pagefault_queue_reset() drops QUEUED entries but
+	 * not the ACTIVE ones which are already in some worker.
+	 * If there are any d->pf.active it means some worker
+	 * will reach eudebug_pagefault_end() and clear
+	 * d->pf.first.
+	 * If there is no active and there if first the
+	 * pagefault was deferred but now queue is cleared
+	 * and no one will reach the first now.
+	 */
+
+	down_read(&xe->usm.lock);
+	xa_for_each(&xe->usm.asid_to_vm, i, vm) {
+		vm = xe_vm_get(vm);
+
+		d = xe_eudebug_get_nolock_with_discovery(vm->xef);
+		if (!d) {
+			xe_vm_put(vm);
+			continue;
+		}
+
+		down_write(&vm->lock);
+		mutex_lock(&xe->eudebug.lock);
+		if (!d->pf.active && d->pf.first &&
+		    d->pf.first->q->gt == gt &&
+		    d->pf.first->q->vm == vm) {
+
+			epf = d->pf.first;
+
+			d->pf.first = NULL;
+			d->pf.asid = 0;
+
+			if (d->pf.fence) {
+				dma_fence_signal(d->pf.fence);
+				dma_fence_put(d->pf.fence);
+				d->pf.fence = NULL;
+			}
+		}
+		mutex_unlock(&xe->eudebug.lock);
+		up_write(&vm->lock);
+
+		if (epf)
+			break;
+
+		xe_vm_put(vm);
+		xe_eudebug_put(d);
+	}
+	up_read(&xe->usm.lock);
+
+	if (epf) {
+		eu_dbg(d, "pagefault 0x%016llx canceled due to GT reset\n",
+		       epf->fault.addr);
+		xe_pm_runtime_get(xe);
+		down_write(&vm->lock);
+		destroy_pagefault(epf);
+		up_write(&vm->lock);
+		xe_pm_runtime_put(xe);
+		xe_vm_put(vm);
+		xe_eudebug_disconnect(d, -EAGAIN);
+		xe_eudebug_put(d);
+		xe_eudebug_attention_poll_start(xe);
+	}
+}
+
+void xe_eudebug_pagefault_gt_init(struct xe_gt *gt)
+{
+	INIT_WORK(&gt->eudebug.gt_reset_work,
+		  eudebug_pagefault_gt_reset_work);
+}
+
+void xe_eudebug_pagefault_reset(struct xe_device *xe,
+				struct xe_gt *gt)
+{
+	if (!xe_eudebug_is_enabled(xe))
+		return;
+
+	queue_work(xe->eudebug.pagefault_reset_wq, &gt->eudebug.gt_reset_work);
+}
+
 bool xe_eudebug_pagefault_creatable(struct xe_gt *gt, struct xe_vm *vm)
 {
 	struct xe_exec_queue *q;
diff --git a/drivers/gpu/drm/xe/xe_eudebug_pagefault.h b/drivers/gpu/drm/xe/xe_eudebug_pagefault.h
index 424bb9679..ce7eca30d 100644
--- a/drivers/gpu/drm/xe/xe_eudebug_pagefault.h
+++ b/drivers/gpu/drm/xe/xe_eudebug_pagefault.h
@@ -8,18 +8,23 @@
 
 #include <linux/types.h>
 
+struct xe_device;
 struct xe_eudebug;
 struct xe_gt;
 struct xe_pagefault;
 struct xe_eudebug_pagefault;
 struct xe_vm;
+struct xe_vma;
 struct xe_file;
 
 #if IS_ENABLED(CONFIG_DRM_XE_EUDEBUG)
 void xe_eudebug_pagefault_fini(struct xe_eudebug *d);
 int xe_eudebug_handle_pagefaults(struct xe_gt *gt);
 
-int xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_pagefault *pf);
+void xe_eudebug_pagefault_gt_init(struct xe_gt *gt);
+
+int xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_vma *vma,
+			       struct xe_pagefault *pf);
 struct xe_vma *xe_eudebug_create_vma(struct xe_vm *vm, struct xe_pagefault *pf);
 void xe_eudebug_pagefault_end(void *private, u32 asid, int err);
 /*
@@ -38,6 +43,8 @@ void xe_eudebug_pagefault_signal(struct xe_eudebug *d);
 
 bool xe_eudebug_pagefault_creatable(struct xe_gt *gt, struct xe_vm *vm);
 
+void xe_eudebug_pagefault_reset(struct xe_device *xe, struct xe_gt *gt);
+
 #else
 #include <linux/errno.h>
 #include <linux/err.h>
@@ -51,8 +58,12 @@ static inline int xe_eudebug_handle_pagefaults(struct xe_gt *gt)
 	return -EOPNOTSUPP;
 }
 
+static inline void xe_eudebug_pagefault_gt_init(struct xe_gt *gt)
+{
+}
+
 static inline int
-xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_pagefault *pf)
+xe_eudebug_pagefault_start(struct xe_vm *vm, struct xe_vma *vma, struct xe_pagefault *pf)
 {
 	return -EOPNOTSUPP;
 }
@@ -86,6 +97,12 @@ static inline bool xe_eudebug_pagefault_creatable(struct xe_gt *gt,
 {
 	return false;
 }
+
+static inline void xe_eudebug_pagefault_reset(struct xe_device *xe,
+					      struct xe_gt *gt)
+{
+}
+
 #endif
 
 #endif /* _XE_EUDEBUG_PAGEFAULT_H_ */
diff --git a/drivers/gpu/drm/xe/xe_eudebug_types.h b/drivers/gpu/drm/xe/xe_eudebug_types.h
index 61fc3342f..b4e6bc12d 100644
--- a/drivers/gpu/drm/xe/xe_eudebug_types.h
+++ b/drivers/gpu/drm/xe/xe_eudebug_types.h
@@ -496,6 +496,11 @@ struct xe_eudebug_pagefault {
 	 * Shall be removed when pagefault is handled.
 	 */
 	u64 null_vma_addr;
+	/**
+	 * @svm_vma: After handling SVM pagefault original mapping shall be restored.
+	 * The svm_vma is the vma which was removed from original SVM mapping.
+	 */
+	struct xe_vma *svm_vma;
 };
 
 #endif /* _XE_EUDEBUG_TYPES_H_ */
diff --git a/drivers/gpu/drm/xe/xe_gt.c b/drivers/gpu/drm/xe/xe_gt.c
index 478e04703..a25c863a1 100644
--- a/drivers/gpu/drm/xe/xe_gt.c
+++ b/drivers/gpu/drm/xe/xe_gt.c
@@ -24,6 +24,7 @@
 #include "xe_debugfs.h"
 #include "xe_device.h"
 #include "xe_eu_stall.h"
+#include "xe_eudebug_pagefault.h"
 #include "xe_exec_queue.h"
 #include "xe_execlist.h"
 #include "xe_force_wake.h"
@@ -723,6 +724,7 @@ int xe_gt_init(struct xe_gt *gt)
 	int i;
 
 	INIT_WORK(&gt->reset.worker, gt_reset_worker);
+	xe_eudebug_pagefault_gt_init(gt);
 
 	for (i = 0; i < XE_ENGINE_CLASS_MAX; ++i) {
 		gt->ring_ops[i] = xe_ring_ops_get(gt, i);
diff --git a/drivers/gpu/drm/xe/xe_gt_types.h b/drivers/gpu/drm/xe/xe_gt_types.h
index 628911346..28158c9e9 100644
--- a/drivers/gpu/drm/xe/xe_gt_types.h
+++ b/drivers/gpu/drm/xe/xe_gt_types.h
@@ -419,6 +419,18 @@ struct xe_gt {
 
 	/** @eu_stall: EU stall counters subsystem per gt info */
 	struct xe_eu_stall_gt *eu_stall;
+
+#if IS_ENABLED(CONFIG_DRM_XE_EUDEBUG)
+	/**
+	 * @eudebug: Worker to finalize eudebug pagefault on gt reset.
+	 */
+	struct {
+		/**
+		 * @eudebug.gt_reset_work: eudebug reset work for that gt.
+		 */
+		struct work_struct gt_reset_work;
+	} eudebug;
+#endif
 };
 
 #endif
diff --git a/drivers/gpu/drm/xe/xe_guc_pagefault.c b/drivers/gpu/drm/xe/xe_guc_pagefault.c
index e65e4962a..c4d75c489 100644
--- a/drivers/gpu/drm/xe/xe_guc_pagefault.c
+++ b/drivers/gpu/drm/xe/xe_guc_pagefault.c
@@ -11,12 +11,13 @@
 #include "xe_hw_engine.h"
 #include "xe_pagefault.h"
 #include "xe_pagefault_types.h"
+#include "xe_eudebug_pagefault.h"
 
 #define XE_GUC_PAGEFAULT_FLUSH_PERIOD	BIT(4)	/* Sixteen */
 
 static void guc_ack_fault_begin(void *private)
 {
-	struct xe_guc *guc = private;
+	struct xe_guc *guc = xe_eudebug_pagefault_get_private(private);
 
 	xe_guc_ct_lock(&guc->ct);
 
@@ -53,7 +54,7 @@ static void guc_ack_fault(struct xe_pagefault *pf, int err)
 		FIELD_PREP(PFR_ENG_CLASS, engine_class) |
 		FIELD_PREP(PFR_PDATA, pdata),
 	};
-	struct xe_guc *guc = pf->producer.private;
+	struct xe_guc *guc = xe_eudebug_pagefault_get_private(pf->producer.private);
 	bool write_only = guc->pagefault_ack_counter++ &
 		(XE_GUC_PAGEFAULT_FLUSH_PERIOD - 1);
 
@@ -61,13 +62,14 @@ static void guc_ack_fault(struct xe_pagefault *pf, int err)
 			      write_only);
 }
 
-static void guc_ack_fault_end(void *private)
+static void guc_ack_fault_end(void *private, u32 asid, int err)
 {
-	struct xe_guc *guc = private;
+	struct xe_guc *guc = xe_eudebug_pagefault_get_private(private);
 
 	if ((guc->pagefault_ack_counter & (XE_GUC_PAGEFAULT_FLUSH_PERIOD - 1)) != 1)
 		xe_guc_ct_send_flush(&guc->ct);
 	xe_guc_ct_unlock(&guc->ct);
+	xe_eudebug_pagefault_end(private, asid, err);
 }
 
 static void xe_guc_pagefault_print(struct xe_pagefault *pf,
diff --git a/drivers/gpu/drm/xe/xe_pagefault.c b/drivers/gpu/drm/xe/xe_pagefault.c
index ba6c3f381..b0b6aa58e 100644
--- a/drivers/gpu/drm/xe/xe_pagefault.c
+++ b/drivers/gpu/drm/xe/xe_pagefault.c
@@ -10,6 +10,7 @@
 
 #include "xe_bo.h"
 #include "xe_device.h"
+#include "xe_eudebug_pagefault.h"
 #include "xe_gt_printk.h"
 #include "xe_gt_types.h"
 #include "xe_gt_stats.h"
@@ -247,9 +248,51 @@ static int xe_pagefault_service(struct xe_pagefault *pf)
 
 	vma = xe_vm_find_vma_by_addr(vm, xe_pagefault_addr(pf));
 	if (!vma) {
+#if IS_ENABLED(CONFIG_DRM_XE_EUDEBUG)
+retry_eudebug_pf:
+#endif
 		err = -EINVAL;
-		xe_pagefault_set_error(pf, XE_PAGEFAULT_ERROR_VMA_NOT_FOUND);
-		goto unlock_vm;
+		if (!xe_eudebug_pagefault_start(vm, vma, pf)) {
+			/*
+			 * This section is reachable only when eudebug is connected.
+			 * It's rather safe since page faults hitting same vma
+			 * are chained. Only the first one will be fully processed.
+			 *
+			 * For SVM retry case we're holding vm write lock
+			 */
+			if (!vma) {
+				up_read(&vm->lock);
+				down_write(&vm->lock);
+				/* The lock was dropped, re-check the vma */
+				vma = xe_vm_find_vma_by_addr(vm, xe_pagefault_addr(pf));
+				if (vma) {
+					err = -EINVAL;
+					downgrade_write(&vm->lock);
+					goto unlock_vm;
+				}
+			}
+			vma = xe_eudebug_create_vma(vm, pf);
+			downgrade_write(&vm->lock);
+			if (IS_ERR(vma)) {
+				err = PTR_ERR(vma);
+				vma = NULL;
+			}
+		} else if (vma) {
+			/*
+			 * xe_eudebug_pagefault_start() failed on svm retry.
+			 * Need to insert again vma.
+			 * Write-lock is still held so no error expected.
+			 */
+			xe_vm_insert_vma(vm, vma);
+			vma = NULL;
+			downgrade_write(&vm->lock);
+		}
+		if (!vma) {
+			xe_pagefault_set_error(pf, XE_PAGEFAULT_ERROR_VMA_NOT_FOUND);
+			goto unlock_vm;
+		}
+	} else {
+		xe_eudebug_pagefault_set_private(pf, vm);
 	}
 
 	if (xe_vma_read_only(vma) &&
@@ -261,11 +304,52 @@ static int xe_pagefault_service(struct xe_pagefault *pf)
 
 	atomic = xe_pagefault_access_is_atomic(pf->consumer.access_type);
 
-	if (xe_vma_is_cpu_addr_mirror(vma))
+	if (xe_vma_is_cpu_addr_mirror(vma)) {
 		err = xe_svm_handle_pagefault(vm, vma, pf, gt,
 					      xe_pagefault_addr(pf), atomic);
-	else
+#if IS_ENABLED(CONFIG_DRM_XE_EUDEBUG)
+		/*
+		 * If err is -ENOENT, it means that the cpu-address-space-mirrored
+		 * xe vma exists, but there is no mm vma allocated in
+		 * the CPU address space. This indicates that no memory has been
+		 * allocated in the CPU address space.
+		 */
+		if (err == -ENOENT &&
+		    !xe_vm_is_closed_or_banned(vm) &&
+		    xe_eudebug_pagefault_creatable(gt, vm)) {
+			u32 page_size = vm->flags & XE_VM_FLAG_64K ? SZ_64K : SZ_4K;
+			u64 fault_addr = ALIGN_DOWN(xe_pagefault_addr(pf), page_size);
+
+			/*
+			 * This section is reachable only when eudebug is connected.
+			 *
+			 * Up until now there was a read-lock. Now entering write-lock.
+			 * Under new lock a vma split may occur. This means that before
+			 * entering write-lock a origin vma could be already split.
+			 * Update the vma to get current one before subtracting.
+			 */
+			up_read(&vm->lock);
+			down_write(&vm->lock);
+			vma = xe_vm_find_vma_by_addr(vm, xe_pagefault_addr(pf));
+			if (vma && xe_vma_is_cpu_addr_mirror(vma))
+				vma = xe_vm_svm_vma_subtract(vm, vma,
+							     fault_addr,
+							     fault_addr + page_size);
+			else
+				vma = ERR_PTR(-EINVAL);
+
+			if (IS_ERR(vma)) {
+				downgrade_write(&vm->lock);
+				err = PTR_ERR(vma);
+			} else {
+				/* keep the write lock for vma */
+				goto retry_eudebug_pf;
+			}
+		}
+#endif
+	} else {
 		err = xe_pagefault_handle_vma(gt, vma, pf, atomic);
+	}
 
 unlock_vm:
 	up_read(&vm->lock);
@@ -620,7 +704,7 @@ static void xe_pagefault_queue_work(struct work_struct *w)
 
 	while (xe_pagefault_queue_pop(pf_queue, &pf, pf_work->id)) {
 		const struct xe_pagefault_ops *ops = pf->producer.ops;
-		void *private = pf->producer.private;
+		void *private;
 		struct xe_gt *gt = pf->gt;
 		u32 asid = FIELD_GET(XE_PAGEFAULT_ASID_MASK, pf->consumer.id);
 		int err = 0;
@@ -629,6 +713,7 @@ static void xe_pagefault_queue_work(struct work_struct *w)
 		/* Last fault same address, ack immediately */
 		if (xe_pagefault_match(pf, cache_start, cache_end, cache_asid)) {
 			xe_gt_stats_incr(gt, XE_GT_STATS_ID_LAST_PAGEFAULT_COUNT, 1);
+			xe_eudebug_pagefault_set_private(pf, NULL);
 			goto ack_fault;
 		}
 
@@ -653,6 +738,12 @@ static void xe_pagefault_queue_work(struct work_struct *w)
 		}
 
 ack_fault:
+		/*
+		 * set private after xe_pagefault_service() since eudebug could swap
+		 * the pf->producer.private field. Also needed when cache was hit.
+		 */
+		private = pf->producer.private;
+
 		xe_assert(xe, pf->consumer.alloc_state ==
 			  XE_PAGEFAULT_ALLOC_STATE_ACTIVE);
 		xe_assert(xe, pf == pf_work->cache.pf);
@@ -692,7 +783,7 @@ ack_fault:
 
 			spin_unlock_irq(&pf_queue->lock);
 		}
-		ops->ack_fault_end(private);
+		ops->ack_fault_end(private, asid, err);
 
 		if (time_after(jiffies, threshold)) {
 			queue_work(xe->usm.pagefault_wq, w);
@@ -852,6 +943,7 @@ static void xe_pagefault_queue_reset(struct xe_device *xe, struct xe_gt *gt,
 void xe_pagefault_reset(struct xe_device *xe, struct xe_gt *gt)
 {
 	xe_pagefault_queue_reset(xe, gt, &xe->usm.pf_queue);
+	xe_eudebug_pagefault_reset(xe, gt);
 }
 
 /*
diff --git a/drivers/gpu/drm/xe/xe_pagefault_types.h b/drivers/gpu/drm/xe/xe_pagefault_types.h
index 96b382ba8..8b9da5da4 100644
--- a/drivers/gpu/drm/xe/xe_pagefault_types.h
+++ b/drivers/gpu/drm/xe/xe_pagefault_types.h
@@ -136,10 +136,12 @@ struct xe_pagefault_ops {
 	/**
 	 * @ack_fault_end: Ack fault end
 	 * @private: producer private data
+	 * @asid: address space ID
+	 * @err: Error state of fault
 	 *
 	 * Page fault producer ends acknowledgment from the consumer.
 	 */
-	void (*ack_fault_end)(void *private);
+	void (*ack_fault_end)(void *private, u32 asid, int err);
 	/**
 	 * @print: Print pagefault info
 	 * @pf: Page fault
-- 
2.43.0

